src/transformStateWithClones.js

```diff
@@ -7,7 +7,46 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear': {
+        currentState = {};
+        break;
+      }
+
+      case 'addProperties': {
+        const stateCopy = { ...currentState, ...action.extraData };
+
+        currentState = stateCopy;
+        break;
+      }
+
+      case 'removeProperties': {
+        const stateCopy = Object.keys(currentState).reduce((newState, key) => {
+          if (!action.keysToRemove.includes(key)) {
+            newState[key] = currentState[key];
+          }
+
+          return newState;
+        }, {});
+
+        currentState = stateCopy;
+        break;
+      }
+
+      default: {
+        throw new Error(`Unknown action type: ${action.type}`);
+      }
+    }
+
+    stateHistory.push(currentState);
```

Comment:
The `stateHistory.push(currentState);` statement should be placed outside of the `switch` block, at the end of each loop cycle, after the current action is processed. This ensures that the state is recorded after all modifications for the current action are complete, as per the additional prompt rules.